plutus-chain-index/app/Marconi/Logging.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Marconi.Logging (logging) where
+
+import Cardano.Api (Block (Block), BlockHeader (BlockHeader), BlockInMode (BlockInMode), CardanoMode,
+                    ChainPoint (ChainPoint), ChainTip (ChainTip), SlotNo (SlotNo, unSlotNo))
+import Cardano.BM.Trace (Trace, logInfo)
+import Control.Monad (when)
+import Data.IORef (IORef, modifyIORef', newIORef, readIORef)
+import Data.Text (Text)
+import Data.Time (NominalDiffTime, UTCTime, defaultTimeLocale, diffUTCTime, formatTime, getCurrentTime,
+                  getCurrentTimeZone, utcToZonedTime)
+import Marconi.Orphans ()
+import Plutus.Streaming (ChainSyncEvent (RollBackward, RollForward))
+import Prettyprinter (Pretty (pretty), defaultLayoutOptions, layoutPretty, (<+>))
+import Prettyprinter.Render.Text (renderStrict)
+import Streaming (Of, Stream, effect)
+import Streaming.Prelude qualified as S
+import Text.Printf (printf)
+
+data SyncStats = SyncStats
+  { -- | Number of applied blocks since last message
+    syncStatsAppliedBlocks :: !Int,
+    -- | Number of rollbacks
+    syncStatsLastRollback  :: !(Maybe UTCTime),
+    -- | Timestamp of last printed message
+    syncStatsLastMessage   :: !(Maybe UTCTime)
+  }
+
+data SyncState = NotSynchronising | Synchronising SlotNo SlotNo Double | Synchronised
+  deriving (Show)
+
+logging ::
+  Trace IO Text ->
+  Stream (Of (ChainSyncEvent (BlockInMode CardanoMode))) IO r ->
+  Stream (Of (ChainSyncEvent (BlockInMode CardanoMode))) IO r
+logging tracer s = effect $ do
+  stats <- newIORef (SyncStats 0 Nothing Nothing)
+  return $ S.chain (update stats) s
+  where
+    minSecondsBetweenMsg :: NominalDiffTime
+    minSecondsBetweenMsg = 10
+
+    update :: IORef SyncStats -> ChainSyncEvent (BlockInMode CardanoMode) -> IO ()
+    update statsRef (RollForward bim ct) = do
+      let cp = case bim of (BlockInMode (Block (BlockHeader slotNo hash _blockNo) _txs) _eim) -> ChainPoint slotNo hash
```

Comment:
I don't know what's best practice, but should we do `_` instead of `_blockNo`? It reduces line length

Reply:
I tend to prefer `_blockNo` because it might look confusing with many ignored parameters (e.g. `BlockInMode (Block (BlockHeader slotNo hash _) _) _`). I'll keep your comment in mind next time.